File: js/arli/src/env.js
/**
 * Environment for arli (JavaScript port) — maps symbols to values.
 */

export class Environment {
  constructor(parent = null, name = 'global') {
    this.parent = parent;
    this.name = name;
    this._bindings = {};
  }

  define(name, value) {
    this._bindings[name] = value;
    return value;
  }

  lookup(name) {
    if (Object.prototype.hasOwnProperty.call(this._bindings, name)) return this._bindings[name];
    if (this.parent !== null) return this.parent.lookup(name);
    return null;
  }

  get(name) {
    const val = this.lookup(name);
    if (val === null || val === undefined) throw new ReferenceError(`Undefined symbol: ${name}`);
    return val;
  }

  set(name, value) {
    if (Object.prototype.hasOwnProperty.call(this._bindings, name)) { this._bindings[name] = value; return value; }
    if (this.parent !== null) return this.parent.set(name, value);
    throw new ReferenceError(`Cannot set! undefined symbol: ${name}`);
  }

  has(name) {
    if (Object.prototype.hasOwnProperty.call(this._bindings, name)) return true;
    if (this.parent !== null) return this.parent.has(name);
    return false;
  }

  extend(name = 'block') { return new Environment(this, name); }

  toString() { return `<Env ${this.name}: ${JSON.stringify(Object.keys(this._bindings))}>`; }
}

File: go/hya/env.go
package main

import "fmt"

// ---------------------------------------------------------------------------
// Environment — lexical scoping with parent chain
// ---------------------------------------------------------------------------

type Environment struct {
	parent  *Environment
	name    string
	bindings map[string]HyaValue
}

func NewEnvironment(parent *Environment, name string) *Environment {
	return &Environment{
		parent:   parent,
		name:     name,
		bindings: make(map[string]HyaValue),
	}
}

func (env *Environment) Define(name string, value HyaValue) HyaValue {
	env.bindings[name] = value
	return value
}

func (env *Environment) Lookup(name string) (HyaValue, bool) {
	if v, ok := env.bindings[name]; ok {
		return v, true
	}
	if env.parent != nil {
		return env.parent.Lookup(name)
	}
	return nil, false
}

func (env *Environment) Get(name string) (HyaValue, error) {
	v, ok := env.Lookup(name)
	if !ok {
		return nil, fmt.Errorf("undefined symbol: %s", name)
	}
	return v, nil
}

func (env *Environment) Set(name string, value HyaValue) error {
	if _, ok := env.bindings[name]; ok {
		env.bindings[name] = value
		return nil
	}
	if env.parent != nil {
		return env.parent.Set(name, value)
	}
	return fmt.Errorf("cannot set! undefined symbol: %s", name)
}
